### DISCUSSION TITLE
Pip doesn't set PYTHONPATH correctly in PEP517 backend

### DISCUSSION THREAD

**Discussion by brandon-leapyear** [Homebrew/discussions#3821] [Everyday usage]
# Minimal repro

First, set up the project:

```bash
mkdir -p /tmp/foo && cd /tmp/foo
echo 'import requests' > /tmp/foo/backend.py
cat > /tmp/foo/pyproject.toml <<EOF
[build-system]
requires = ["requests"]
build-backend = "backend"
backend-path = ["."]
EOF
```

I'm just using `requests` here to represent some third-party package that pip needs to install to run the backend. I'm actually using `poetry-core` and it's failing when importing `poetry.core.masonry.api`, but I'm trying to use a more common library here to prove that it's not due to poetry.

Building this project with the `pip` bundled by Homebrew directly fails:
```console
$ /usr/local/opt/python@3.10/bin/python3.10 -m pip wheel .
Processing /tmp/foo
  Installing build dependencies ... done
  Getting requirements to build wheel ... done
ERROR: Exception:
Traceback (most recent call last):
  ...
  File "/tmp/foo/backend.py", line 2, in <module>
    import requests
ModuleNotFoundError: No module named 'requests'
```

But doing it with a pip in a virtualenv resolves the import correctly and gets a different error:
```console
$ /usr/local/opt/python@3.10/bin/python3.10 -m venv venv

$ readlink venv/bin/python3.10
/usr/local/opt/python@3.10/bin/python3.10

$ venv/bin/python3.10 -m pip wheel .
Processing /tmp/foo
  Installing build dependencies ... done
  Getting requirements to build wheel ... done
  Preparing metadata (pyproject.toml) ... error
  error: subprocess-exited-with-error
  
  × Preparing metadata (pyproject.toml) did not run successfully.
  │ exit code: 1
  ╰─> [15 lines of output]
      ...
      AttributeError: module 'backend' has no attribute 'build_wheel'
      [end of output]
```

Printing out `sys.path` in `backend.py` (and adding `-v` to the `pip wheel` command), I noticed that the two calls have different paths:
```python
# plain pip
[
    '$PWD',
    '/usr/local/lib/python3.10/site-packages/pip/_vendor/pep517/in_process',
    '$TMPDIR/pip-build-env-t8xgo00t/site',
    '/usr/local/Cellar/python@3.10/3.10.8/Frameworks/Python.framework/Versions/3.10/lib/python310.zip',
    '/usr/local/Cellar/python@3.10/3.10.8/Frameworks/Python.framework/Versions/3.10/lib/python3.10',
    '/usr/local/Cellar/python@3.10/3.10.8/Frameworks/Python.framework/Versions/3.10/lib/python3.10/lib-dynload',
    '/usr/local/lib/python3.10/site-packages',
]

# venv pip
[
    '$PWD',
    '/tmp/foo/venv/lib/python3.10/site-packages/pip/_vendor/pep517/in_process',
    '$TMPDIR/pip-build-env-h2nywc6a/site',
    '/usr/local/Cellar/python@3.10/3.10.8/Frameworks/Python.framework/Versions/3.10/lib/python310.zip',
    '/usr/local/Cellar/python@3.10/3.10.8/Frameworks/Python.framework/Versions/3.10/lib/python3.10',
    '/usr/local/Cellar/python@3.10/3.10.8/Frameworks/Python.framework/Versions/3.10/lib/python3.10/lib-dynload',
    '$TMPDIR/pip-build-env-h2nywc6a/overlay/lib/python3.10/site-packages',
    '$TMPDIR/pip-build-env-h2nywc6a/normal/lib/python3.10/site-packages',
]
```
Specifically, the plain pip call doesn't include the `overlay` or `normal` directories, which is where pip installs dependencies listed in `build-backend`. The only thing I can think of is that Homebrew installs pip differently, in a way that creating a virtualenv installs a fresh unbroken pip.

# Notes

I recently upgraded to Monterey, not sure if this is related. I also completely uninstalled/removed Homebrew + reinstalled, but it's still happening on a fresh install.

**Comment by SMillerDev** [comment]
I'm having a hard time figuring out what the specific Homebrew bug is here.

**Reply by brandon-leapyear** [reply]
I'm also unsure what in Homebrew is causing it, but I'm fairly certain it's a Homebrew bug because `/usr/bin/python3 -m pip wheel .` works. That is,
1. Built-in Python: works
2. Homebrew Python: fails
3. Venv Python created by Homebrew Python: works

@SMillerDev are you able to repro this at all?

**Reply by SMillerDev** [reply]
Seeing how it's completely different python versions I'd suggest filing a big upstream first since they know their system best.

**Reply by brandon-leapyear** [reply]
I tried it with the same version as Homebrew. Updated list:

1. Built-in Python (3.9): works
2. Homebrew Python (3.10): fails
3. Venv Python created by Homebrew Python (3.10): works
4. Python installed from upstream (3.10): works
